### ISSUE TITLE
RFC: `nim c` output path

### ISSUE THREAD

**Issue by euantorano** [nim-lang/Nim#7142]
Currently, invoking `nim c` to compile Nim code writes the generated binary into the directory that the source file was in.

We were discussing this on IRC the other day, and [Araq said I should open an issue](https://irclogs.nim-lang.org/24-01-2018.html#17:10:28) to discuss this point.

To give some background, see the below examples:

**Example 1 - simple flat file structure**

Source tree:

```
.
└── hello.nim
```

After running `nim c hello.nim`:

```
.
├── hello
├── hello.nim
└── nimcache
    ├── hello.c
    ├── hello.json
    ├── hello.o
    ├── stdlib_system.c
    └── stdlib_system.o
```

**Example 2 - file structure with a `src` subdirectory**

Source tree:

```
.
└── src
    └── hello.nim
```

After running `him c src/hello.nim`:

```
.
└── src
    ├── hello
    ├── hello.nim
    └── nimcache
        ├── hello.c
        ├── hello.json
        ├── hello.o
        ├── stdlib_system.c
        └── stdlib_system.o
```

See that both the `nimcache` directory and the `hello` executable were written to the `src` path, along with the source.

## The problem

Writing binary and cache files into the subdirectory that code is compiled from feels unintuitive to me, and is different behaviour from almost every other tool I can think of.

This is also a problem for Nimble, and is what leads to the suggestion that binary projects with modules name the root module with a `pkg` suffix  (such as [`nimblepkg`](https://github.com/nim-lang/nimble/tree/master/src/nimblepkg)) otherwise you would end up trying to write a binary with the same name as a directory and hitting an error.

## The expectation

If I invoke `nim c foo.nim`, I expect a `foo` executable to be written to my current working directory, resulting in `./foo` (or `./foo.exe`). I would also expect the `nimcache` to be written to the current directory, resulting in a tree like the below from the above second example:

```
.
├── hello
├── nimcache
│   ├── hello.c
│   ├── hello.json
│   ├── hello.o
│   ├── stdlib_system.c
│   └── stdlib_system.o
└── src
    └── hello.nim
```

In flat directory structures with more modules like the below, there would still be an error on platforms that have no suffix for executable files (eg: everything except Windows):

```
.
├── hello
│   └── somemod.nim
├── hello.nim
```

In that case, the `-o:` argument would need to be used to specify an output file path, but @dom96 suggested that Nim could be smart here and append some kind of suffix to the binary to prevent conflicts.

## Other tools

In order to show why I find this odd, I thought it best to include some examples of what some other tools do. I've picked a few basic examples which are included below.

1. **D (DMD):**

	Source tree:
	
	```
	.
	└── src
	  └── hello.d
	```
	
	After running `dmd src/hello.d`:
	
	```
	.
	├── hello
	├── hello.o
	└── src
		└── hello.d
	```
	
2. **C (clang):**
	
	Source tree:
	
	```
	.
	└── src
		└── hello.c
	```
	
	After running `clang src/hello.c`:
	
	```
	.
	├── a.out
	└── src
		└── hello.c
	```
	
3. **Crystal:**

	Source tree:
	
	```
	.
	└── src
		└── hello.cr
	```
	
	After running `crystal build src/hello.cr`:
	
	```
	.
	├── hello
	├── hello.dwarf
	└── src
		└── hello.cr
	```

**Comment by Araq** [comment]
Using the current working directory is problematic for reliable build setups.

``nim c -r tests/tester`` 

should IMO be equivalent to

``cd tests; nim c -r tester``

The current solution of using addressing relative to the ``tester.nim`` file accomplishes this. We can make Nim produce ``foo.elf`` as the default for a binary on UNIX should a ``foo`` directory already exist.

**Comment by euantorano** [comment]
Can you explain why it's problematic for reliable build setups? It seems to work perfectly fine for every single other compiler available.

**Comment by Araq** [comment]
I have explained why it's problematic. And it doesn't work "perfectly fine" for every other compiler available.

**Comment by dom96** [comment]
@Araq Please elaborate on your explanation. It's not clear what you mean.

**Comment by yglukhov** [comment]
The build scripts can use `--output` and `--nimcache` options easily.

**Comment by Quelklef** [comment]
I personally like the default behavior, even if it's inconsistent with other projects. Perhaps the default behavior could be what @euantorano has suggested, but a terse option (`-rel`?) could be added to allow for the old behavior. It would be much nicer than including `--output` and `--nimcache` each time.

**Comment by euantorano** [comment]
For even more inconsistency, take the following source tree:

```
.
└── src
    └── test.nim
```

Run the command `nim doc2 src/test.nim`.

Guess where `test.html` ends up? 😛 

```
.
├── src
│   └── test.nim
└── test.html
```

**Comment by Araq** [comment]
> For even more inconsistency, take the following source tree...

That only means we need to fix ``nim doc2``. :P

**Comment by siliconvoodoo** [comment]
Araq is right to revise the industry. if something is wrong with everybody it's good to recognize it. Though adding optional -o parameter is needed. and indeed fix doc2 to behave the same.

**Comment by dom96** [comment]
> Araq is right to revise the industry. if something is wrong with everybody it's good to recognize it. 

Sure, but that requires something to be wrong with how everybody is doing it. I still don't understand what is wrong. Can somebody explain?

**Comment by euantorano** [comment]
It's also worth noting that Nimble (`v0.8.10`) seems to output binaries where one would expect:

```
.
├── src
│   └── test.nim
└── test.nimble
```

After `nimble build`:

```
.
├── src
│   └── test.nim
└── test.nimble
└── test.exe
```

**Comment by dom96** [comment]
Nimble outputs binaries beside ``test.nimble`` by default. Whereas ``nim doc`` puts it in the CWD if I'm not mistaken. Of course you can only run ``nimble build`` in the working directory of ``test.nimble`` anyway... so it works out the same.

**Comment by euantorano** [comment]
Yep, and it makes sense to do it that way in my opinion @dom96.

**Comment by narimiran** [comment]
> Currently, invoking nim c to compile Nim code writes the generated binary into the directory that the source file was in.

This was changed recently (v0.19.0) and compiled files are now in `~/.cache/nim/filename`.

Should this be closed now?

**Comment by euantorano** [comment]
I don’t believe so. If running a standard compile does place the binary there (eg: well away from anywhere anybody would expect), then I’d say it’s got worse.
